File: Documentation/5.4/Samples/python/ClientApi/Session/LoadingEntities.py
from typing import Optional, Union, Dict, Callable, Type, List, TypeVar

from ravendb import IncludeBuilder, LoaderWithInclude, ConditionalLoadResult
from ravendb.infrastructure.entities import User
from ravendb.infrastructure.orders import Employee, Product, Supplier

from examples_base import ExamplesBase

_T = TypeVar("_T")


class LoadingEntities(ExamplesBase):
    def setUp(self):
        super().setUp()
        with self.embedded_server.get_document_store("LoadingEntitiesXY").open_session() as session:
            session.store(Product(supplier="suppliers/1"), "products/1")
            session.save_changes()

    class Foo:
        # region loading_entities_1_0
        def load(
            self,
            key_or_keys: Union[List[str], str],
            object_type: Optional[Type[_T]] = None,
            includes: Callable[[IncludeBuilder], None] = None,
        ) -> Union[Dict[str, _T], _T]:
            ...

        # endregion
        # region loading_entities_2_0
        def include(self, path: str) -> LoaderWithInclude:
            ...

        # endregion

        # region loading_entities_3_0
        def load(
            self,
            key_or_keys: Union[List[str], str],  # <- List of ids
            object_type: Optional[Type[_T]] = None,
            includes: Callable[[IncludeBuilder], None] = None,
        ) -> Union[Dict[str, _T], _T]:
            ...

        # endregion
        # region loading_entities_4_0
        def load_starting_with(
            self,
            id_prefix: str,
            object_type: Optional[Type[_T]] = None,
            matches: Optional[str] = None,
            start: Optional[int] = None,
            page_size: Optional[int] = None,
            exclude: Optional[str] = None,
            start_after: Optional[str] = None,
        ) -> List[_T]:
            ...

        def load_starting_with_into_stream(
            self,
            id_prefix: str,
            output: bytes,
            matches: str = None,
            start: int = 0,
            page_size: int = 25,
            exclude: str = None,
            start_after: str = None,
        ):
            ...

    # endregion
    # region loading_entities_5_0
    # unsupported, will be supported from 5.4 client release (https://pypi.org/project/ravendb/)
    # endregion
    # region loading_entities_6_0
    def is_loaded(self, key: str) -> bool:
        ...

    # endregion

    # region loading_entities_7_0
    def conditional_load(self, key: str, change_vector: str, object_type: Type[_T] = None) -> ConditionalLoadResult[_T]:
        ...

    # endregion

    def test_loading_entities_xy(self):
        with self.embedded_server.get_document_store("LoadingEntitiesXY") as store:
            with store.open_session() as session:
                # region loading_entities_1_1

                employee = session.load("employees/1", Employee)

                # endregion

            with store.open_session() as session:
                # region loading_entities_2_1

                # loading 'products/1'
                # including document found in 'supplier' property
                products_by_key = session.include("supplier").load(Product, "products/1")
                product = products_by_key["products/1"]

                supplier = session.load(product.supplier)  # this will not make server call

                # endregion

            with store.open_session() as session:
                # region loading_entities_2_2

                # loading 'products/1'
                # including document found in 'Supplier' property
                products_by_key = session.include("Supplier").load(Product, "products/1")
                product = products_by_key["products/1"]

                supplier = session.load(product.supplier, Supplier)
                # endregion

            with store.open_session() as session:
                # region loading_entities_3_1
                employees = session.load(["employees/1", "employees/2", "employees/3"], Employee)
                # endregion

            with store.open_session() as session:
                # region loading_entities_4_1
                # return up to 128 entities with Id that starts with 'employees'
                result = session.advanced.load_starting_with("employees/", Employee, None, 0, 128)
                # endregion

            with store.open_session() as session:
                # region loading_entities_4_2
                # return up to 128 entities with Id that starts with 'employees'
                # and rest of the key begins with "1" or "2" e.g. employees/10, employees/25
                result = session.advanced.load_starting_with("employees/", Employee, "1*|2*", 0, 128)
                # endregion

            with store.open_session() as session:
                # region loading_entities_5_1
                # unsupported, will be supported from 5.4 client release (https://pypi.org/project/ravendb/)
                # endregion
                ...
            with store.open_session() as session:
                session.advanced.load_starting_with_into_stream = lambda x, y: None
                # region loading_entities_5_2
                stream_bytes = bytes(b"My loaded documents will go here -> ")
                session.advanced.load_starting_with_into_stream("employees/", stream_bytes)
                # endregion

            with store.open_session() as session:
                # region loading_entities_6_1
                is_loaded = session.advanced.is_loaded("employees/1")  # False
                employee = session.load("employees/1")
                is_loaded = session.advanced.is_loaded("employees/1")  # True
                # endregion

            # region loading_entities_7_1
            change_vector: Optional[str] = None
            user = User(name="Bob")

            with store.open_session() as session:
                session.store(user, "users/1")
                session.save_changes()

                change_vector = session.advanced.get_change_vector_for(user)

            # Now session which does not track our User entity
            with store.open_session() as session:
                # The given change vector matches
                # the server-side change vector
                # Does not load the document
                result1 = session.advanced.conditional_load("users/1", change_vector)

                # Modify the document
                user.name = "Bob Smith"
                session.store(user)
                session.save_changes()

                # Change vectors do not natch
                # Loads the document
                result2 = session.advanced.conditional_load("user/1", change_vector)
            # endregion

File: Documentation/5.4/Samples/php/ClientApi/session/LoadingEntities.php
<?php

namespace RavenDB\Samples\ClientApi\Session;

use RavenDB\Documents\Session\ConditionalLoadResult;
use RavenDB\Documents\Session\Loaders\LoaderWithIncludeInterface;
use RavenDB\Samples\Infrastructure\DocumentStoreHolder;
use RavenDB\Samples\Infrastructure\Entity\User;
use RavenDB\Samples\Infrastructure\Orders\Employee;
use RavenDB\Samples\Infrastructure\Orders\Product;
use RavenDB\Samples\Infrastructure\Orders\Supplier;
use RavenDB\Type\ObjectArray;
use RavenDB\Type\StringArray;

interface FooInterface {
    # region loading_entities_1_0
    public function load(?string $className, ?string $id): ?object;
    # endregion
}

interface FooInterface2 {
    # region loading_entities_2_0
    function include(?string $path): LoaderWithIncludeInterface;

    public function load(string $className, array $ids): ObjectArray;
    public function load(string $className, StringArray $ids): ObjectArray;

    public function load(string $className, string $id): ?object;
    # endregion
}

interface FooInterface3 {
    # region loading_entities_3_0
    public function load(string $className, array $ids): ObjectArray;
    public function load(string $className, StringArray $ids): ObjectArray;
    # endregion

    # region loading_entities_4_0
    public function loadStartingWith(
        string $className,
        ?string $idPrefix,
        ?string $matches = null,
        int $start = 0,
        int $pageSize = 25,
        ?string $exclude = null,
        ?string $startAfter = null
    ): ObjectArray;

    # endregion

    # region loading_entities_6_0
    function isLoaded(string $id): bool;
    # endregion

    # region loading_entities_7_0
    function conditionalLoad(?string $className, ?string $id, ?string $changeVector): ConditionalLoadResult;
    # endregion
}

class LoadingEntities
{
    public function testSamples(): void
    {
        $store = DocumentStoreHolder::getStore();
        try {
            $session = $store->openSession();
            try {
                # region loading_entities_1_1

                /** @var Employee $employee */
                $employee = $session->load(Employee::class, "employees/1");
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region loading_entities_2_1
                // loading 'products/1'
                // including document found in 'supplier' property

                /** @var Product $product */
                $product = $session
                    ->include("Supplier")
                    ->load(Product::class, "products/1");

                $supplier = $session->load(Supplier::class, $product->getSupplier()); // this will not make server call
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region loading_entities_2_2
                // loading 'products/1'
                // including document found in 'supplier' property
                /** @var Product $product */
                $product = $session
                    ->include("Supplier")
                    ->load(Product::class, "products/1");

                $supplier = $session->load(Supplier::class, $product->getSupplier()); // this will not make server call
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region loading_entities_3_1
                $employees = $session->load(Employee::class, "employees/1", "employees/2", "employees/3");
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region loading_entities_4_1
                // return up to 128 entities with Id that starts with 'employees'
                $result = $session
                    ->advanced()
                    ->loadStartingWith(Employee::class, "employees/", null, 0, 128);
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region loading_entities_4_2
                // return up to 128 entities with Id that starts with 'employees/'
                // and rest of the key begins with "1" or "2" e.g. employees/10, employees/25
                $result = $session
                    ->advanced()
                    ->loadStartingWith(Employee::class, "employees/", "1*|2*", 0, 128);
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region loading_entities_6_1
                $isLoaded = $session->advanced()->isLoaded("employees/1"); //false
                $employee = $session->load(Employee::class, "employees/1");
                $isLoaded = $session->advanced()->isLoaded("employees/1"); // true
                # endregion
            } finally {
                $session->close();
            }

            # region loading_entities_7_1
            $session = $store->openSession();
            try {
                $changeVector = "";
                $user = new User("Bob");

                $session->store($user, "users/1");
                $session->saveChanges();

                $changeVector = $session->advanced()->getChangeVectorFor($user);
            } finally {
                $session->close();
            }

            $user = new User("Bob");
            $changeVector = "a";

            $session = $store->openSession();
            try {
                // New session which does not track our User entity

                // The given change vector matches
                // the server-side change vector
                // Does not load the document
                $result1 = $session->advanced()
                    ->conditionalLoad(User::class, "users/1", $changeVector);

                // Modify the document
                $user->setName("Bob Smith");
                $session->store($user);
                $session->saveChanges();

                // Change vectors do not match
                // Loads the document
                $result2 = $session->advanced()
                    ->conditionalLoad(User::class, "users/1", $changeVector);
            } finally {
                $session->close();
            }
            # endregion
        } finally {
            $store->close();
        }
    }
}
